tests/apps/core/test_core_api_room_group_access.py

```diff
@@ -12,148 +13,331 @@
-        administrator = UserFactory()
-        group = GroupFactory(members=[administrator])
-        access = RoomGroupAccessFactory(room__groups=[(group, True)])
+        user = UserFactory()
+        jwt_token = AccessToken.for_user(user)
 
-        jwt_token = AccessToken.for_user(administrator)
+        public_room = RoomFactory(is_public=True, users=[(user, "owner")])
+        RoomGroupAccessFactory(room=public_room)
+        RoomGroupAccessFactory(room=public_room, role="member")
+        RoomGroupAccessFactory(room=public_room, role="administrator")
+
+        private_room = RoomFactory(is_public=False, users=[(user, "owner")])
+        RoomGroupAccessFactory(room=private_room)
+        RoomGroupAccessFactory(room=private_room, role="member")
+        RoomGroupAccessFactory(room=private_room, role="administrator")
 
         response = self.client.get(
-            f"/api/rooms/{access.room.id!s}/groups/{access.group.id!s}/",
+            "/api/room-group-accesses/",
             HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
         )
         self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.json()), 6)
+
+    # Retrieve
+
+    def test_api_room_group_accesses_retrieve_anonymous(self):
+        """
+        Anonymous users should not be allowed to retrieve a room group access.
+        """
+        access = RoomGroupAccessFactory()
+        response = self.client.get(
+            f"/api/room-group-accesses/{access.id!s}/",
+        )
 
+        self.assertEqual(response.status_code, 401)
         self.assertEqual(
-            response.json(),
-            {
-                "id": str(access.id),
-                "group": str(access.group.id),
-                "room": str(access.room.id),
-                "is_administrator": access.is_administrator,
-            },
+            response.json(), {"detail": "Authentication credentials were not provided."}
         )
 
+    def test_api_room_group_accesses_retrieve_authenticated(self):
+        """
+        Authenticated users should not be allowed to retrieve a room group access for
+        a room to which they are not related, be it public or private.
+        """
+        user = UserFactory()
+        jwt_token = AccessToken.for_user(user)
+
+        for is_public in [True, False]:
+            room = RoomFactory(is_public=is_public)
+            self.assertEqual(len(GroupRoleChoices.choices), 2)
+
+            for role, _name in GroupRoleChoices.choices:
+                access = RoomGroupAccessFactory(room=room, role=role)
+                response = self.client.get(
+                    f"/api/room-group-accesses/{access.id!s}/",
+                    HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+                )
+                self.assertEqual(response.status_code, 403)
+                self.assertEqual(
+                    response.json(),
+                    {"detail": "You do not have permission to perform this action."},
+                )
+
```

Comment:
For complicated factorization yes he was right! I'm just doing simple loops here. Did you get it straightaway or was it disturbing. If it was hard to understand in one glance, i can split it in several tests but it will take a lot of place. Compromises :man_shrugging: